File: src/main/java/com/nanaios/polygonal_tech/capability/interfaces/ILongFluidTank.java
package com.nanaios.polygonal_tech.capability.interfaces;

import com.nanaios.polygonal_tech.fluids.base.LongFluidStack;
import com.nanaios.polygonal_tech.util.MathUtil;
import net.minecraft.nbt.CompoundTag;
import net.minecraftforge.fluids.FluidStack;
import net.minecraftforge.fluids.IFluidTank;
import net.minecraftforge.fluids.capability.IFluidHandler.FluidAction;
import org.jetbrains.annotations.NotNull;

/// long型の流体量を扱うためのインターフェース
public interface ILongFluidTank extends IFluidTank ,ICapability{
    String NBT_STORED_LONG_FLUID_STACK = "stored_long_fluid_stack";

    /// タンク内の流体量をlong型で取得します。
    long getFluidLongAmount();

    /// タンクの最大容量をlong型で取得します。
    long getLongCapacity();

    /// 流体を受け取ります。
    ///
    /// @param resource 受け取る流体の種類と量をLongFluidStackで指定します。
    /// @param action   受け取りのシミュレーションを行うかどうかを指定します。
    /// @return 実際に受け取った流体の量をlong型で返します。
    long fillLong(LongFluidStack resource, FluidAction action);

    /// タンク内の流体をLongFluidStackで取得します。
    ///
    /// @return タンク内の流体をLongFluidStackで返します。
    @Override
    @NotNull
    LongFluidStack getFluid();

    /// 流体を抽出します。
    ///
    /// @param maxDrain 抽出する流体の最大量をlong型で指定します。
    /// @param action   抽出のシミュレーションを行うかどうかを指定します。
    /// @return 実際に抽出された流体の種類と量をLongFluidStackで返します。
    @NotNull
    LongFluidStack drain(long maxDrain, FluidAction action);

    /// 流体を抽出します。
    ///
    /// @param resource 抽出する流体の種類と量をLongFluidStackで指定します。
    /// @param action   抽出のシミュレーションを行うかどうかを指定します。
    /// @return 実際に抽出された流体の種類と量をLongFluidStackで返します。
    @NotNull
    LongFluidStack drain(LongFluidStack resource, FluidAction action);

    @Override
    default CompoundTag serializeNBT() {
        CompoundTag tag = new CompoundTag();
        tag.put(NBT_STORED_LONG_FLUID_STACK, getFluid().writeToNBT(tag));
        return tag;
    }

    @Override
    default void deserializeNBT(CompoundTag nbt) {
        if (nbt.contains(NBT_STORED_LONG_FLUID_STACK)) {
            LongFluidStack fluidStack = LongFluidStack.loadLongFluidStackFromNBT(nbt.getCompound(NBT_STORED_LONG_FLUID_STACK));
            fillLong(fluidStack, FluidAction.EXECUTE);
        }
    }

    @NotNull
    default LongFluidStack drain(int maxDrain, FluidAction action) {
        return drain((long) maxDrain, action);
    }

    default int getFluidAmount() {
        return MathUtil.longToInt(getFluidLongAmount());
    }

    @Override
    @NotNull
    default LongFluidStack drain(FluidStack resource, FluidAction action) {
        return drain(LongFluidStack.from(resource), action);
    }

    @Override
    default int getCapacity() {
        return MathUtil.longToInt(getLongCapacity());
    }

    @Override
    default int fill(FluidStack resource, FluidAction action) {
        return MathUtil.longToInt(fillLong(LongFluidStack.from(resource), action));
    }
}

File: src/main/kotlin/com/nanaios/polygonal_tech/core/capability/fluid/ILongFluidTank.kt
package com.nanaios.polygonal_tech.core.capability.fluid

import com.nanaios.polygonal_tech.core.capability.ICapability
import com.nanaios.polygonal_tech.core.fluid.LongFluidStack
import com.nanaios.polygonal_tech.core.fluid.toLongFluidStack
import com.nanaios.polygonal_tech.core.util.roundInt
import net.minecraftforge.fluids.FluidStack
import net.minecraftforge.fluids.IFluidTank
import net.minecraftforge.fluids.capability.IFluidHandler.FluidAction

/**
 * 基本的な[IFluidTank]（流体タンク）のインターフェースを拡張し、
 * [Int.MAX_VALUE]を超える大容量の流体（[LongFluidStack]）を直接管理・操作し、
 * かつネットワーク同期（[ICapability]）に対応させることを目的としたインターフェース。
 */
interface ILongFluidTank: IFluidTank, ICapability {
    /** タンク内で現在保持している大容量流体データ */
    val longFluid: LongFluidStack
    /** 現在保持している流体の[Long]型数量 */
    val longFluidAmount: Long
    /** このタンクが保持できる流体の最大容量（[Long]型） */
    val longCapability: Long

    /**
     * 投入されようとしている大容量流体（[LongFluidStack]）が、このタンクの受け入れ条件を満たしているかを判定することを目的とする。
     *
     * @param stack 判定・シミュレート対象の流体
     * @return 受け入れ可能なら[true]
     */
    fun isFluidValid(stack: LongFluidStack): Boolean

    /**
     * 大容量流体を受け入れてタンクに充填することを目的とするメソッド。
     *
     * @param resource 充填しようとする流体
     * @param action 実際の充填を行うか、シミュレートかで動作を分ける[FluidAction]
     * @return 実際に受け入れた（または受け入れ可能な）[Long]型の流体量
     */
    fun fill(resource: LongFluidStack, action: FluidAction): Long

    /**
     * タンク内から指定した最大[Long]量の流体を抽出することを目的とするメソッド。
     *
     * @param maxDrain 抽出したい最大流体量
     * @param action 実際の抽出を行うか、シミュレートかで動作を分ける[FluidAction]
     * @return 抽出された流体情報を持つ[LongFluidStack]
     */
    fun drain(maxDrain: Long, action: FluidAction): LongFluidStack

    /**
     * 特定の流体の種類と量を指定してタンク内から抽出することを目的とするメソッド。
     *
     * @param resource 抽出したい流体の条件を指定する[LongFluidStack]
     * @param action シミュレートか実行かを示す[FluidAction]
     * @return 抽出された[LongFluidStack]
     */
    fun drain(resource: LongFluidStack, action: FluidAction): LongFluidStack

    override fun getFluid() = longFluid.toFluidStack()
    override fun getFluidAmount() = longFluidAmount.roundInt()
    override fun getCapacity() = longCapability.roundInt()
    override fun isFluidValid(stack: FluidStack) = isFluidValid(stack.toLongFluidStack())
    override fun fill(resource: FluidStack, action: FluidAction) = fill(resource.toLongFluidStack(), action).roundInt()
    override fun drain(maxDrain: Int, action: FluidAction) = drain(maxDrain.toLong(), action).toFluidStack()
    override fun drain(resource: FluidStack, action: FluidAction) = drain(resource.toLongFluidStack(), action).toFluidStack()
}